receipt_listener.go

```diff
@@ -74,149 +82,154 @@ func (l *ReceiptListener) Run(ctx context.Context) error {
+		logs := txLogs[txHash]
 
-				l.log.Debug().
-					Str("tx", txHash.Hex()).
-					Str("meta-tx", string(metaTxnID)).
-					Msgf("Found succeed meta-tx")
+		// metaTxnReceipts: meta transaction receipts indexed by meta transaction ID
+		// receipt: gets the current receipt for a given meta transaction ID
 
-				// Failed transactions have the TxFailed topic and the data begins with the metaTxInd
-			} else if len(txLog.Topics) == 1 && txLog.Topics[0] == TxFailedEventSig && len(txLog.Data) >= 32 {
-				status = MetaTxnFailed
-				metaTxnID = MetaTxnID(common.Bytes2Hex(txLog.Data[:32]))
-
-				l.log.Debug().
-					Str("tx", txHash.Hex()).
-					Str("meta-tx", string(metaTxnID)).
-					Msgf("Found failed meta-tx")
-			} else {
-				continue // unknown, skip
+		metaTxnReceipts := map[common.Hash]*ReceiptResult{}
+		receipt := func(metaTxnID common.Hash) *ReceiptResult {
+			result := metaTxnReceipts[metaTxnID]
+			if result == nil {
+				result = &ReceiptResult{MetaTxnID: MetaTxnID(metaTxnID.Hex()[2:])}
+				metaTxnReceipts[metaTxnID] = result
 			}
+			return result
+		}
+
+		// assuming block.Logs is sorted
+		for _, log := range logs {
+			if len(log.Topics) == 0 && len(log.Data) == 32 {
+				// possible TxExecuted event
 
-			result := &ReceiptResult{
-				MetaTxnID:  metaTxnID,
-				Status:     status,
-				TxnReceipt: tx,
+				r := receipt(common.BytesToHash(log.Data))
+				r.Results = append(r.Results, &MetaTxnResult{
+					Status: MetaTxnExecuted,
+				})
+			} else if len(log.Topics) == 1 && log.Topics[0] == TxFailedEventSig {
+				// definite TxFailed event
+
+				metaTxnID, reason, err := DecodeTxFailedEvent(log)
+				if err != nil {
+					l.log.Err(err).Msgf("unable to decode TxFailed event: topics=%v data=%v", log.Topics, log.Data)
+					continue
+				}
+
+				r := receipt(metaTxnID)
+				r.Results = append(r.Results, &MetaTxnResult{
+					Status: MetaTxnFailed,
+					Reason: reason,
+				})
 			}
+		}
 
-			// Add found result to block of receipts
-			blockOfReceipts = append(blockOfReceipts, result)
+		receipts[txHash] = make([]*ReceiptResult, 0, len(metaTxnReceipts))
+		for _, receipt := range metaTxnReceipts {
+			receipts[txHash] = append(receipts[txHash], receipt)
 		}
 	}
 
-	// Nothing to record, skipping.
-	if len(blockOfReceipts) == 0 {
-		return nil
+	for txHash, txReceipts := range receipts {
+		l.handleReceipts(txHash, txReceipts)
+	}
+}
+
+func (l *ReceiptListener) handleReceipts(txHash common.Hash, txReceipts []*ReceiptResult) {
+	if len(txReceipts) == 0 {
+		return
 	}
 
-	// Publish to subscribers
-	l.mu.Lock()
-	defer l.mu.Unlock()
-
-	for _, result := range blockOfReceipts {
-		for _, sub := range l.subscribers {
-			select {
-			case <-sub.done:
-			case sub.ch <- result:
-			case <-time.After(2 * time.Second):
-				l.log.Warn().Msgf("channel publisher is blocked by a slow subscriber")
+	// handle receipts in an independent goroutine so that node failures don't stall the block handler
+	go func() {
```

Comment:
the issue here is that each time we call `handleReceipts` we will be spinning up a new goroutine, which is inexpensive, but we have an unbounded number of parallel remove http calls we will be making via TransactionReceipt within this goroutine. Instead we should have a `MaxConcurrentFetchReceipts = 30` .. which is cool that we will be doing parallel queries :) this will make this even faster too, but we should limit this.

one way to bound this with the goroutine is to have a token bucket to rate-limit parallelism..

```go
// put this in the constructor or somewhere of ReceiptListener to fill up the bucket
limiter := chan(struct{}, MaxConcurrentFetchReceipts)
for i := 0; i < MaxConcurrentFetchReceipts; i++ {
  limiter <- struct{}
}

// then.. inside of handleReceipts
go func() {
  <-l.limiter
  defer l.limiter <- struct{}{}
  
  // rest of code goes here..
}()
```